### PR TITLE
[Improvement]: Element deletion asynchronous queue

### LINKED ISSUE #9758: [Feature]: Element deletion asynchronous queue
### Feature description

In object storage setups asset deletion is quite an expensive task. A better approach would be to mark assets as deleted only an later on delete the files in the filesystem via an asynchronous queue. Instead of just creating a deleted flag a better approach would be to delete the entry from the assets table and save it into either another table or to some kind of queue (Symfony messenger?).

One problem might be that the public URL is still reachable until the queue is executed. Maybe it could be a good idea to rename them to some kind of random name during deletion synchronously as renaming is much faster then deleting the file.

### PR DESCRIPTION
**Pull request by kingjia90**
## Changes in this pull request  
Resolves [#9758](https://github.com/pimcore/pimcore/issues/9758)

## Additional info
Only renaming for local publicly accessible files, on cloud storage i am avoiding it as it may result in an extra copy and delete operation.

### CODE REVIEW

File: lib/Messenger/Handler/AssetDeleteHandler.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        $storage = Storage::get('asset');
+
+        try {
+            if ($message->getType() === 'folder') {
+                $storage->deleteDirectory($message->getFullPath());
+            } else {
+                $storage->delete($message->getFullPath());
+            }
+        } catch (Exception $e) {
+            Logger::err(sprintf('Problem deleting the asset phyisical file with Id: %s and fullpath: %s, reason: %s',
```
Comment (jdreesen) [inline lib/Messenger/Handler/AssetDeleteHandler.php]:
```suggestion
            Logger::err(sprintf('Problem deleting the asset physical file with ID: %s and fullpath: %s, reason: %s',
```

File: lib/Messenger/AssetDeleteMessage.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        return $this->fullPath;
+    }
+
+    public function getId(): int
+    {
+        return $this->id;
+    }
+
+    // For "type" is meant whether it is a folder or a file here
+    public function getType(): string
```
Comment (jdreesen) [inline lib/Messenger/AssetDeleteMessage.php]: If it's only `file` or `folder` why not use a boolean `isFolder` instead?

Comment (kingjia90) [reply lib/Messenger/AssetDeleteMessage.php]: thank you! 
yes, that's a good idea, at least is less ambiguous

**Comment by kingjia90** [comment]
Mhhh i re-read the initial issue 

> In object storage setups asset deletion is quite an expensive task.

Then would need to apply the queue to the cloud storage as well since it wasn't as i hoped them to be.

What about to change the visibility/unix permissions of the files to avoid being accessed by direct url (instead of renaming)?

File: lib/Messenger/Handler/AssetDeleteHandler.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        // Clear thumbnails and asset cache
+        foreach (['thumbnail', 'asset_cache'] as $storageName) {
+            $storage = Storage::get($storageName);
+            $storage->deleteDirectory($message->getFullPath());
+        }
+
+        $db = Db::get();
+        $db->delete('assets_image_thumbnail_cache', ['cid' => $message->getId()]);
```
Comment (kingjia90) [inline lib/Messenger/Handler/AssetDeleteHandler.php:62]: this probably should be synchronously deleted in the model 🤔

**Comment by kingjia90** [comment]
Closing, see https://github.com/pimcore/pimcore/issues/9758#issuecomment-2303983886

**Pull request closed without merging**